File: leoz-node/src/main/java/org/deku/leo2/node/rest/ExceptionMapper.java
package org.deku.leo2.node.rest;

import com.fasterxml.jackson.core.JsonLocation;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leo2.rest.entities.v1.Error;

import javax.inject.Named;
import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.Provider;
import java.util.Optional;

/**
 * Created by masc on 21.04.15.
 */
@Named
@Provider
public class ExceptionMapper implements javax.ws.rs.ext.ExceptionMapper<Exception> {
    Log mLog = LogFactory.getLog(this.getClass());

    @Override
    public javax.ws.rs.core.Response toResponse(Exception e) {
        Error result;

        if (e instanceof ServiceException) {
            ServiceException se = (ServiceException) e;
            result = new Error(se.getResponse().getStatus(), se.getErrorCode().ordinal(), se.getMessage());

        } else if (e instanceof WebApplicationException) {
            //region WebApplicationException
            WebApplicationException we = (WebApplicationException) e;
            result = new Error(we.getResponse().getStatus(), we);
            //endregion

        } else if (e instanceof JsonMappingException) {
            //region JsonMappingException
            JsonMappingException jm = (JsonMappingException) e;
            String locationMessage = String.join(".", (Iterable)jm.getPath().stream().map(p -> p.getFieldName())::iterator);

            result = new Error(Response.Status.BAD_REQUEST.getStatusCode(),
                    String.format("JSON mapping error [%s]: %s", locationMessage,
                            (jm.getCause() != null) ? jm.getCause().getMessage() : "unknown"));
            //endregion

        } else if (e instanceof JsonProcessingException) {
            //region JsonProcessingException
            JsonProcessingException je = (JsonProcessingException) e;
            JsonLocation jl = je.getLocation();

            Optional<String> locationMessage = Optional.empty();
            if (jl != null) {
                locationMessage = Optional.of(String.format(" in line %d column %d", jl.getLineNr(), jl.getColumnNr()));
            }

            result = new Error(Response.Status.BAD_REQUEST.getStatusCode(),
                    String.format("JSON processing error%s: %s",
                            locationMessage.orElse(""),
                            je.getOriginalMessage()));
            //endregion

        } else {
            result = new Error(Response.Status.BAD_REQUEST.getStatusCode(), e);
        }

        return Response
                .status(result.getStatus())
                .entity(result)
                .type(MediaType.APPLICATION_JSON_TYPE)
                .build();
    }
}

File: leoz-node/src/main/java/org/deku/leoz/node/rest/ExceptionMapper.kt
package org.deku.leoz.node.rest

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.JsonMappingException
import org.deku.leoz.service.entity.ServiceError
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.zalando.problem.Status
import org.zalando.problem.ThrowableProblem
import sx.rs.RestProblem
import javax.ws.rs.WebApplicationException
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.ext.Provider

/**
 * Created by masc on 21.04.15.
 */
@Component
@Provider
class ExceptionMapper : javax.ws.rs.ext.ExceptionMapper<Exception> {
    private val log = LoggerFactory.getLogger(this.javaClass)

    override fun toResponse(e: Exception): javax.ws.rs.core.Response {
        val status: Int
        val entity: Any?

        when(e) {
            is WebApplicationException -> return e.response

            is NoSuchElementException -> {
                status = Response.Status.NOT_FOUND.statusCode
                entity = RestProblem(status = Status.NOT_FOUND, detail = e.message)
            }
            is ServiceException -> {
                if (e.entity != null) {
                    entity = e.entity
                }
                else if (e.errorCode != null) {
                    entity = ServiceError(e.errorCode, e.message ?: "")
                } else {
                    entity = null
                }
                status = e.status.statusCode
            }
            is ThrowableProblem -> {
                entity = e
                status = e.status?.statusCode ?: Response.Status.BAD_REQUEST.statusCode

            }
            is JsonMappingException -> {
                //region JsonMappingException
                log.error(e.message)
                val locationMessage = e.path.map { it.fieldName }.joinToString(".")
                val cause = e.cause?.message ?: "unknown"

                status = Response.Status.BAD_REQUEST.statusCode
                entity = RestProblem(status = Status.BAD_REQUEST, detail = "JSON mapping error [${locationMessage}]: ${cause}")
                //endregion
            }
            is JsonProcessingException -> {
                //region JsonProcessingException
                log.error(e.message)
                val jl = e.location

                val locationMessage: String = if (jl != null) " in line ${jl.lineNr} column ${jl.columnNr}" else ""

                status = Response.Status.BAD_REQUEST.statusCode
                entity = RestProblem(status = Status.BAD_REQUEST, detail = "JSON processing error${locationMessage}: ${e.originalMessage}")
                //endregion
            }
            else -> {
                log.error(e.message, e)
                status = Response.Status.BAD_REQUEST.statusCode
                entity = RestProblem(status = Status.BAD_REQUEST, detail = e.message)
            }
        }

        return Response
                .status(status)
                .entity(entity)
                .type(MediaType.APPLICATION_JSON_TYPE)
                .build()
    }
}